README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# BarcodeOverlapAnalyzer

<!-- badges: start -->
<!-- badges: end -->

The goal of BarcodeOverlapAnalyzer is to …

## Installation

You can install the development version of BarcodeOverlapAnalyzer from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("foocheung/golemffs")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(BarcodeOverlapAnalyzer)
## basic example code
```

What is special about using `README.Rmd` instead of just `README.md`?
You can include R chunks like so:

``` r
summary(cars)
#>      speed           dist       
#>  Min.   : 4.0   Min.   :  2.00  
#>  1st Qu.:12.0   1st Qu.: 26.00  
#>  Median :15.0   Median : 36.00  
#>  Mean   :15.4   Mean   : 42.98  
#>  3rd Qu.:19.0   3rd Qu.: 56.00  
#>  Max.   :25.0   Max.   :120.00
```

You’ll still need to render `README.Rmd` regularly, to keep `README.md`
up-to-date. `devtools::build_readme()` is handy for this. You could also
use GitHub Actions to re-render `README.Rmd` every time you push. An
example workflow can be found here:
<https://github.com/r-lib/actions/tree/v1/examples>.

You can also embed plots, for example:

<img src="man/figures/README-pressure-1.png" width="100%" />

In that case, don’t forget to commit and push the resulting figure
files, so they display on GitHub and CRAN.


NEWS.md
# BarcodeOverlapAnalyzer 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/app_ui.R
#' The application User-Interface
#'
#' @param request Internal parameter for `{shiny}`.
#'     DO NOT REMOVE.
#' @import shiny
#' @noRd

app_ui <- function() {
  #tagList(
  # Leave this function for adding external resources
  shinyUI( pageWithSidebar(
    HTML("<CENTER><H2>BarcodeOverlapAnalyzer"),

    tabPanel(
      "",
      sidebarPanel(width = 3,
                   "",
                   mod_dataInput_ui("dataInput_ui")
      )
    ),
    mainPanel(
   #   tabsetPanel(

        mod_dataInput_ui2("dataInput_ui2")


    #  )
    )
  ))
}

#' Add external Resources to the Application
#'
#' This function is internally used to add external
#' resources inside the Shiny application.
#'
#' @import shiny
#' @importFrom golem add_resource_path activate_js favicon bundle_resources
#' @noRd
golem_add_external_resources <- function() {
  add_resource_path(
    "www",
    app_sys("app/www")
  )

  tags$head(
    favicon(),
    bundle_resources(
      path = app_sys("app/www"),
      app_title = "rbarcodes"
    )
    # Add here other external resources
    # for example, you can add shinyalert::useShinyalert()
  )
}


dev/02_dev.R
# Building a Prod-Ready, Robust Shiny Application.
#
# README: each step of the dev files is optional, and you don't have to
# fill every dev scripts before getting started.
# 01_start.R should be filled at start.
# 02_dev.R should be used to keep track of your development during the project.
# 03_deploy.R should be used once you need to deploy your app.
#
#
###################################
#### CURRENT FILE: DEV SCRIPT #####
###################################

# Engineering

## Dependencies ----
## Amend DESCRIPTION with dependencies read from package code parsing
## install.packages('attachment') # if needed.
attachment::att_amend_desc()

## Add modules ----
## Create a module infrastructure in R/
golem::add_module(name = "name_of_module1", with_test = TRUE) # Name of the module
golem::add_module(name = "name_of_module2", with_test = TRUE) # Name of the module

## Add helper functions ----
## Creates fct_* and utils_*
golem::add_fct("helpers", with_test = TRUE)
golem::add_utils("helpers", with_test = TRUE)

## External resources
## Creates .js and .css files at inst/app/www
golem::add_js_file("script")
golem::add_js_handler("handlers")
golem::add_css_file("custom")
golem::add_sass_file("custom")

## Add internal datasets ----
## If you have data in your package
usethis::use_data_raw(name = "my_dataset", open = FALSE)

## Tests ----
## Add one line by test you want to create
usethis::use_test("app")

# Documentation

## Vignette ----
usethis::use_vignette("BarcodeOverlapAnalyzer")
devtools::build_vignettes()

## Code Coverage----
## Set the code coverage service ("codecov" or "coveralls")
usethis::use_coverage()

# Create a summary readme for the testthat subdirectory
covrpage::covrpage()

## CI ----
## Use this part of the script if you need to set up a CI
## service for your application
##
## (You'll need GitHub there)
usethis::use_github()

# GitHub Actions
usethis::use_github_action()
# Chose one of the three
# See https://usethis.r-lib.org/reference/use_github_action.html
usethis::use_github_action_check_release()
usethis::use_github_action_check_standard()
usethis::use_github_action_check_full()
# Add action for PR
usethis::use_github_action_pr_commands()

# Travis CI
usethis::use_travis()
usethis::use_travis_badge()

# AppVeyor
usethis::use_appveyor()
usethis::use_appveyor_badge()

# Circle CI
usethis::use_circleci()
usethis::use_circleci_badge()

# Jenkins
usethis::use_jenkins()

# GitLab CI
usethis::use_gitlab_ci()

# You're now set! ----
# go to dev/03_deploy.R
rstudioapi::navigateToFile("dev/03_deploy.R")


R/app_config.R
#' Access files in the current app
#'
#' NOTE: If you manually change your package name in the DESCRIPTION,
#' don't forget to change it here too, and in the config file.
#' For a safer name change mechanism, use the `golem::set_golem_name()` function.
#'
#' @param ... character vectors, specifying subdirectory and file(s)
#' within your package. The default, none, returns the root of the app.
#'
#' @noRd
app_sys <- function(...) {
  system.file(..., package = "rbarcodes")
}


#' Read App Config
#'
#' @param value Value to retrieve from the config file.
#' @param config GOLEM_CONFIG_ACTIVE value. If unset, R_CONFIG_ACTIVE.
#' If unset, "default".
#' @param use_parent Logical, scan the parent directory for config file.
#' @param file Location of the config file
#'
#' @noRd
get_golem_config <- function(
  value,
  config = Sys.getenv(
    "GOLEM_CONFIG_ACTIVE",
    Sys.getenv(
      "R_CONFIG_ACTIVE",
      "default"
    )
  ),
  use_parent = TRUE,
  # Modify this if your config file is somewhere else
  file = app_sys("golem-config.yml")
) {
  config::get(
    value = value,
    config = config,
    file = file,
    use_parent = use_parent
  )
}


dev/01_start.R
# Building a Prod-Ready, Robust Shiny Application.
#
# README: each step of the dev files is optional, and you don't have to
# fill every dev scripts before getting started.
# 01_start.R should be filled at start.
# 02_dev.R should be used to keep track of your development during the project.
# 03_deploy.R should be used once you need to deploy your app.
#
#
########################################
#### CURRENT FILE: ON START SCRIPT #####
########################################

## Fill the DESCRIPTION ----
## Add meta data about your application
##
## /!\ Note: if you want to change the name of your app during development,
## either re-run this function, call golem::set_golem_name(), or don't forget
## to change the name in the app_sys() function in app_config.R /!\
##
golem::fill_desc(
  pkg_name = "BarcodeOverlapAnalyzer", # The Name of the package containing the App
  pkg_title = "PKG_TITLE", # The Title of the package containing the App
  pkg_description = "PKG_DESC.", # The Description of the package containing the App
  author_first_name = "AUTHOR_FIRST", # Your First Name
  author_last_name = "AUTHOR_LAST", # Your Last Name
  author_email = "AUTHOR@MAIL.COM", # Your Email
  repo_url = NULL, # The URL of the GitHub Repo (optional),
  pkg_version = "0.0.0.9000" # The Version of the package containing the App
)

## Set {golem} options ----
golem::set_golem_options()

## Install the required dev dependencies ----
golem::install_dev_deps()

## Create Common Files ----
## See ?usethis for more information
usethis::use_mit_license("Golem User") # You can set another license here
usethis::use_readme_rmd(open = FALSE)
devtools::build_readme()
# Note that `contact` is required since usethis version 2.1.5
# If your {usethis} version is older, you can remove that param
usethis::use_code_of_conduct(contact = "Golem User")
usethis::use_lifecycle_badge("Experimental")
usethis::use_news_md(open = FALSE)

## Use git ----
usethis::use_git()

## Init Testing Infrastructure ----
## Create a template for tests
golem::use_recommended_tests()

## Favicon ----
# If you want to change the favicon (default is golem's one)
golem::use_favicon() # path = "path/to/ico". Can be an online file.
# golem::remove_favicon() # Uncomment to remove the default favicon

## Add helper functions ----
golem::use_utils_ui(with_test = TRUE)
golem::use_utils_server(with_test = TRUE)

# You're now set! ----

# go to dev/02_dev.R
rstudioapi::navigateToFile("dev/02_dev.R")


R/mod_dataInput.R
#' dataInput UI Function
#'
#' @description A shiny Module.
#'
#' @param id,input,output,session Internal parameters for {shiny}.
#'
#' @noRd
#'
#' @importFrom shiny NS tagList

`%>%` <- dplyr::`%>%`
addResourcePath("d", "inst/app/www/")
mod_dataInput_ui <- function(id){
  ns <- NS(id)

  tagList(
#    mainPanel(
#     tabsetPanel(
    selectizeInput(ns("pathInput"), "Enter Path to Files:", options = list(create = TRUE), c(
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/*/outs/per_sample_outs/*/count/sample_filtered_barcodes.csv" ,
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/GEX_VDJ/*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/GEX_ADT/*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_*_*/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_*_*/outs/filtered_feature_bc_matrix/aggregate_barcodes.csv",
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_BCR_?/outs/cell_barcodes.json",
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_ADT_?/outs/filtered_feature_bc_matrix/barcodes.tsv.gz",
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_TCR_?/outs/cell_barcodes.json",
      "/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_2/RUN/*/outs/per_sample_outs/*/count/sample_filtered_barcodes.csv"

    ), selected = character(0)),
    div(style="display: inline-block;horizontal-align:top; width: 15%;",textInput(ns("sub1"),label="Label1")),
    div(style="display: inline-block;horizontal-align:top; width: 15%;",textInput(ns("sub2"),label="Label2")),
    # withSpinner(actionButton("submitBtn", "Submit"))
    actionButton(ns("submitBtn"), "Submit")
    #)
  )

}


#' dataInput Server Functions
#'
#' @noRd
mod_dataInput_server <- function(id){
  moduleServer( id, function(input, output, session){
    ns <- session$ns

    output$text<-renderText(
      input$pathInput
    )

    read_csv_filename <- function(filename){

      if(xfun::file_ext(filename) == 'json'){
        ret <- as.tibble(fromJSON(filename))
      }
      else{

        ret <- readr::read_csv(filename,col_names = FALSE) %>% dplyr::select(last_col())

      }

      #  ret <- read_csv(filename,col_names = FALSE) %>% select(last_col())

      # HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_GEX_1/outs/filtered_feature_bc_matrix/barcodes.tsv.gz<BR>")

      colnames(ret) = c("bar")
      ret$file <- gsub(".*RUN/(.*)/outs.*", "\\1", filename)
      ret
    }

 #  inc<- observeEvent(input$submitBtn, {
      # withSpinner({


    userFile <- reactive({
      req(input$submitBtn)
      validate(need(input$pathInput !="", ""))
      ddd<-input$pathInput
      input$pathInput
    })



    ##   observeEvent(input$submitBtn, {
    datafile <- reactive({



   req(input$submitBtn)
      isolate({
      #
    #  inFile<-read_csv_filename()
      progress <- shiny::Progress$new()
      # Make sure it closes when we exit this reactive, even if there's an error
      on.exit(progress$close())

      progress$set(message = "Making plot", value = 0)



        path <- userFile()

       #path <- input$pathInput

      temp <- system(paste("ls ", path, sep = ""), intern = TRUE)
      data_list <- lapply(temp, read_csv_filename)
      result <- do.call(rbind, data_list)

      df <- result
      ddf<-result

      df$file =gsub(input$sub1,input$sub2,df$file)

      bar_lists <- split(df$bar, df$file)
  #    aaa<<-input$sub1



      # Initialize a matrix to store the intersection counts
      intersection_counts <- matrix(0, nrow = length(bar_lists), ncol = length(bar_lists),
                                    dimnames = list(names(bar_lists), names(bar_lists)))

      # Calculate intersections and populate the matrix
      for (i in 1:length(bar_lists)) {
        for (j in i:length(bar_lists)) {
          intersection_counts[i, j] <- length(intersect(bar_lists[[i]], bar_lists[[j]]))
          intersection_counts[j, i] <- intersection_counts[i, j]
        }
        progress$inc(1/length(bar_lists), detail = paste("Doing part", i))
        Sys.sleep(0.1)
        }



      return(
list(
          "intersection_counts" = intersection_counts
        )
)
      })
})





     return(list
            (subtn = reactive({input$submitBtn}),
              df = datafile )
     )



  })
}



R/mod_dataInput2.R
#' dataInput UI Function
#'
#' @description A shiny Module.
#'
#' @param id,input,output,session Internal parameters for {shiny}.
#'
#' @noRd
#'
#' @importFrom shiny NS tagList

`%>%` <- dplyr::`%>%`
addResourcePath("d", "inst/app/www/")
mod_dataInput_ui2 <- function(id){
  ns <- NS(id)

  tagList(


    tabsetPanel(
      tabPanel("Intersection Counts Heatmap",

HTML("<BR>"),

               div(style="display: inline-block;vertical-align:top; margin-left: 80px; width: 15%;",selectInput(ns("norm"), "Normalization", c("normalize", "percentize", "scale"), selected = "normalize")),
               div(style="display: inline-block;horizontal-align:top; width: 15%;",sliderInput(ns("width"), "Plot Width:",
                                                                                               min = 400, max = 2000, value = 600)) ,
               div(style="display: inline-block;horizontal-align:top; width: 15%;",sliderInput(ns("height"), "Plot Height:",
                                                                                               min = 400, max = 2000, value = 600)),
               plotly::plotlyOutput(ns("heatmap")),
               textOutput(ns("text")),
      ),
      tabPanel("Counts Table", DT::DTOutput(ns("table"))),
      tabPanel("Urls",
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/GEX_VDJ/outs/filtered_feature_bc_matrix/barcodes.tsv.gz<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/GEX_ADT/outs/filtered_feature_bc_matrix/barcodes.tsv.gz<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/23_306_3*/outs/per_sample_outs/*/count/sample_filtered_barcodes.csv<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_GEX_?/outs/filtered_feature_bc_matrix/barcodes.tsv.gz<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_BCR_?/outs/cell_barcodes.json<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_ADT_?/outs/filtered_feature_bc_matrix/barcodes.tsv.gz<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_TCR_?/outs/cell_barcodes.json<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_2/RUN/*/outs/per_sample_outs/*/count/sample_filtered_barcodes.csv<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/*/outs/per_sample_outs/*/count/sample_filtered_barcodes.csv<BR>"),
               HTML("/Volumes/CHI/PROJECTS_Archive/2022_CHI_PROPOSALS/Manthiram_Covid-tonsil_CHI-306/NEW_DATA_9_sept_2023/230908_A00941_1368_AHKVGVDSX2/RUN_23_306_3/RUN/2OUT_*_?/outs/cell_barcodes.json<BR>")

      ))
  )
}


#' @noRd
#' @export
#' @keywords internal
#'

mod_dataInput_server2 <- function(input, output, session, file){  #,batches,sim){
  ns <- session$ns

    output$heatmap <-  plotly::renderPlotly({
  req(file$subtn())
        in1<-file$df()$intersection_counts
    in2<<-in1

    ## if(input$sub1){
    #  df$file =gsub(input$sub1,input$sub2,df$file)
    # }

      if (input$norm == "normalize") {
        heatmaply::heatmaply(heatmaply::normalize(in1), cellnote_textposition="middle center", cellnote = in1, plot_method = "ggplot",fontsize = 7, width = input$width, height = input$height )
      }
      else if(input$norm == "percentize"){
        heatmaply::heatmaply(heatmaply::percentize(in1),  cellnote_textposition="middle center", cellnote = in1,plot_method = "ggplot",fontsize = 7, width = input$width, height = input$height )
      }
      else if(input$norm == "scale"){
        #  heatmaply::heatmaply(scale="column",(t(as.matrix(as.data.frame(d)))), plot_method="plotly",labRow=rownames(d) ,row_side_colors =rownames(d), width = input$width, height = input$height )
        heatmaply::heatmaply(scale="none",log(in1 +1), cellnote_textposition="middle center", cellnote = in1 +1,  plot_method = "ggplot",fontsize = 7, width = input$width, height = input$height )

      }



  })



}
